tutorials/autotvm/tune_tensor_core_matmul.py

```diff
@@ -0,0 +1,219 @@
+if len(sys.argv) >= 5:
+  dtype = sys.argv[4]
+if len(sys.argv) >= 6:
+  layout = sys.argv[5]
+
+print ("M=%d, N=%d, K=%d, dtype=%s, layout=%s" % (M, N, L, dtype, layout))
+
+task = autotvm.task.create(test_gemm_nn, args=(N, L, M, dtype, layout), target='cuda')
+print(task.config_space)
+
+logging.getLogger('autotvm').setLevel(logging.DEBUG)
+logging.getLogger('autotvm').addHandler(logging.StreamHandler(sys.stdout))
+
+measure_option = autotvm.measure_option(
+    builder='local',
+    runner=autotvm.LocalRunner(number=5))
+
+tuner = autotvm.tuner.XGBTuner(task)
+with tvm.build_config():
+    tuner.tune(n_trial=1000,
+               measure_option=measure_option,
+               callbacks=[autotvm.callback.log_to_file('matmul.log')])
+
+dispatch_context = autotvm.apply_history_best("matmul.log")
+best_config = dispatch_context.query(task.target, task.workload)
+print("\nBest config:")
+print(best_config)
+with autotvm.apply_history_best('matmul.log'):
+    with tvm.target.create("cuda"):
+        with tvm.build_config():
+            s, arg_bufs = test_gemm_nn(N, L, M, dtype, layout)
+            print(tvm.lower(s, arg_bufs, simple_mode=True))
+            func = tvm.build(s, arg_bufs)
+dev_module = func.imported_modules[0]
+print(dev_module.get_source())
+
+# check correctness
+if (layout == "NN"):
+  shape_a = (N, L)
+  shape_b = (L, M)
+elif (layout == "NT"):
+  shape_a = (L, N)
+  shape_b = (L, M)
+elif (layout == "TN"):
+  shape_a = (N, L)
+  shape_b = (M, L)
+elif (layout == "TT"):
+  shape_a = (L, N)
+  shape_b = (M, L)
+
+a_np = None
+b_np = None
+c_np = None
+c_np_type = None
+if dtype == 'float16':
+  c_np_type = np.float32
+  a_np = np.random.uniform(size=shape_a).astype(np.float16)
+  b_np = np.random.uniform(size=shape_b).astype(np.float16)
+  if (layout == "NN"):
+    c_np = np.dot(a_np, b_np)
+  elif (layout == "NT"):
+    c_np = np.dot(a_np.T, b_np)
+  elif (layout == "TN"):
+    c_np = np.dot(a_np, b_np.T)
+  elif (layout == "TT"):
+    c_np = np.dot(a_np.T, b_np.T)
+elif dtype == 'int8':
+  c_np_type = np.int32
+  a_np = np.random.randint(low=-128, high=127, size=shape_a).astype(np.int8)
+  b_np = np.random.randint(low=-128, high=127, size=shape_b).astype(np.int8)
+  if (layout == "NN"):
+    c_np = np.dot(a_np.astype(np.int32), b_np.astype(np.int32))
+  elif (layout == "NT"):
+    c_np = np.dot(a_np.astype(np.int32).T, b_np.astype(np.int32))
+  elif (layout == "TN"):
+    c_np = np.dot(a_np.astype(np.int32), b_np.astype(np.int32).T)
+  elif (layout == "TT"):
+    c_np = np.dot(a_np.astype(np.int32).T, b_np.astype(np.int32).T)
+
+ctx = tvm.gpu()
+c_tvm = tvm.nd.array(np.zeros(c_np.shape, dtype=c_np_type), ctx=ctx)
+a_tvm = tvm.nd.array(a_np, ctx=ctx)
+b_tvm = tvm.nd.array(b_np, ctx=ctx)
+func(a_tvm, b_tvm, c_tvm)
+
+tvm.testing.assert_allclose(c_np, c_tvm.asnumpy(), rtol=1e-3)
+
+evaluator = func.time_evaluator(func.entry_name, ctx, number=100)
+print('Time cost of this operator: %f' % evaluator(a_tvm, b_tvm, c_tvm).mean)
```

Comment:
Are the tuned results same ones reflected in the RFC? Is this template flexible enough to achieve good performance for other shapes after tuning?

Reply:
Yes, the perf in the RFC are tuned with this tutorial script. The template is not customized for specific shapes so it should be flexible to apply on other shapes. But we do see poor performance on large shapes, as @Hzfengsy commented above. We found out 2 reasons:
1. Bank conflicts of shared memory, which can be reduced by storage_align as @Hzfengsy suggested.
2. "vthread" in this template is a fixed value of 1 instead of a tunable knob. We do so because the inject_virtual_thread pass does not support intrinsics.
The improvement for the first issue will get updated to this PR soon, but for the second one we are still trying to figure out a solution.

Reply:
With storage_align, now the tuned results are better than the ones in the RFC.